src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -2062,21 +2065,30 @@ ContextCoversJoinRestriction(JoinRestrictionContext *joinRestrictionContext,
 			continue;
 		}
 
+		List *joinRestrictInfoListInTest =
+			joinRestrictionInTest->joinRestrictInfoList;
+		bool joinIsOnTrue = list_length(joinRestrictInfoListInTest) == 0;
+		bool hasOnlyInnerJoin = joinRestrictionContext->hasOnlyInnerJoin;
+		if (!hasOnlyInnerJoin && joinIsOnTrue)
+		{
+			/*
+			 * If join doesn't have a restriction (e.g., ON (true)) and planner
+			 * is aware of at least one non-inner JOIN (e.g., outer/semi joins),
+			 * we should not eliminiate joinRestrictionInTest. It can still be
+			 * useful for detecting not supported outer-join checks even if it
+			 * doesn't help for colocation checks.
+			 */
+			continue;
+		}
+
 		/*
 		 * We check whether the restrictions in joinRestrictionInTest is a subset
 		 * of the restrictions in joinRestrictionInContext in the sense that all the
 		 * restrictions in the latter already exists in the former.
-		 *
-		 * Also, note that list_difference() returns a list that contains all the
```

Comment:
why we remove this comment? Can't we adjust?   I think having at least the following would be nice: ` equal() function is implemented for the struct.`

Reply:
Both we already mention `equal` function in `LeftListIsSubset` and that we use `LeftListIsSubset` instead of `list_difference`, I thought it is already clear what we do here.  What do you think ?